### PR TITLE
mgr/dashboard: Allow viewing and setting pool quotas

### PR DESCRIPTION
**Pull request by bk201**
Allow managing pool quotas in both frontend and backend.

Fixes: https://tracker.ceph.com/issues/36559
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by bk201** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts [hunk 1]
```diff
@@ -136,7 +136,15 @@ export class PoolFormComponent implements OnInit {
           validators: [Validators.required, Validators.min(1)]
         }),
         ecOverwrites: new FormControl(false),
-        compression: compressionForm
+        compression: compressionForm,
+        max_bytes: new FormControl(0, {
+          validators: [Validators.required],
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: By reading the error messages I think you are missing the `min` validator:
```suggestion
          validators: [Validators.required, Validators.min(0)],
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 2]
```diff
@@ -427,6 +427,58 @@
...
+                     type="text"
+                     formControlName="max_bytes"
+                     i18n-placeholder
+                     placeholder="e.g., 10GiB"
+                     defaultUnit="GiB"
+                     cdDimlessBinary>
+              <span class="help-block"
+                    i18n>Specify 0 to disable quota</span>
+              <span class="help-block"
+                    *ngIf="form.showError('max_bytes', formDir, 'required')"
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Think this should reflect the `min` validation or both
```suggestion
                    *ngIf="form.showError('max_bytes', formDir, 'min')"
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 3]
```diff
@@ -427,6 +427,58 @@
...
+            </label>
+            <div class="col-sm-9">
+              <input class="form-control"
+                     id="max_objects"
+                     [min]="0"
+                     name="max_objects"
+                     type="number"
+                     formControlName="max_objects">
+              <span class="help-block"
+                    i18n>Specify 0 to disable quota</span>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: I wonder if we should allow the input to be empty.
Think we allow other places to leave it empty if we don't want to change its default value.

Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Thanks, change behavior of these two new form control.

`quota_max_bytes` and `quota_max_objects` fields are
- not included: when fields are not dirty or user clear the fields
- included: when user enters valid values

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 2 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: removed this error condition since min validator does not apply on text input

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 1 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: removed required error condition and allow empty field.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts [hunk 4]
```diff
@@ -136,7 +136,14 @@ export class PoolFormComponent implements OnInit {
           validators: [Validators.required, Validators.min(1)]
         }),
         ecOverwrites: new FormControl(false),
-        compression: compressionForm
+        compression: compressionForm,
+        max_bytes: new FormControl(0, {
+          updateOn: 'blur'
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: I thinks this is already the default and can be removed.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts [hunk 5]
```diff
@@ -136,7 +136,14 @@ export class PoolFormComponent implements OnInit {
...
+        }),
+        max_objects: new FormControl(0, {
+          validators: [Validators.min(0)],
+          updateOn: 'blur'
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: I thinks this is already the default and can be removed.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 6]
```diff
@@ -423,6 +423,57 @@
...
+        <!-- Quotas -->
+        <div>
+          <legend i18n>Quotas</legend>
+
+          <!-- Max Bytes -->
+          <div class="form-group"
+               [ngClass]="{'has-error': form.showError('max_bytes', formDir)}">
+            <label class="control-label col-sm-3"
+                   for="max_bytes">
+              <ng-container i18n>Max bytes</ng-container>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: I would suggest adding the info here and change the text to the following:
```suggestion
              <ng-container i18n>Max bytes</ng-container>
              <cd-helper>
                <span i18n>Leave it blank or specify 0 to disable this quota.</span>
                <br>
                <span i18n>A valid quota should be greater then 0.</span>
              </cd-helper>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 7]
```diff
@@ -423,6 +423,57 @@
...
+                     id="max_bytes"
+                     name="max_bytes"
+                     type="text"
+                     formControlName="max_bytes"
+                     i18n-placeholder
+                     placeholder="e.g., 10GiB"
+                     defaultUnit="GiB"
+                     cdDimlessBinary>
+              <span class="help-block"
+                    i18n>Specify 0 to disable quota</span>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Since this is not related to validation I would remove this and add it in the label inside a `cd-helper`, as suggested in a previous comment.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 8]
```diff
@@ -423,6 +423,57 @@
...
+              <span class="help-block"
+                    i18n>The value should be greater or equal to 0</span>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Since this is not related to validation I would remove this and add it in the label inside a `cd-helper`, as suggested in a previous comment.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 3 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Since this is not related to validation I would remove this and add it beside the label, the same way that I used for the `max_bytes`.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 5 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: can we keep min validator because max_object is a number field on page?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 6 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: When creating a new pool, it's natural not to set any quota if a user left the fields empty.
When editing a existed pool, if a user left the fields empty, should we keep previous quota values of pool? Or just disable quota (set to 0) if the field are empty?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 5 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: I was referring only to the `updateOn` line, the validator should stay.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 6 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: I think if the user left the field empty, we should set the value to the default.
In this case default is disabled, right?

Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: It make sense, since the user clear the field. Thanks!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 9]
```diff
@@ -423,6 +423,57 @@
...
+          <div class="form-group"
+               [ngClass]="{'has-error': form.showError('max_objects', formDir)}">
+            <label class="control-label col-sm-3"
+                   for="max_objects">
+              <ng-container i18n>Max objects</ng-container>
+            </label>
+            <div class="col-sm-9">
+              <input class="form-control"
+                     id="max_objects"
+                     [min]="0"
```
[the remark this answers is not in the document]
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Thanks for spotting this.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 4 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: seems `updateOn` is default to `change`?
https://angular.io/api/forms/AbstractControl#updateOn

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: I had only tested how the input was working and the value was only being updated on blur.

But since `cdDimlessBinary` still works without changing this setting and we don't have any `valueChanges` linked to both this `FormControls` I would suggest to try to keep the changes to a minimum and still remove this.

Hope I'm not missing anything, let me know what you think.

Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: Thanks!
will remove, no valueChanges are linked to these two control.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 10]
```diff
@@ -423,6 +423,61 @@
...
+          <legend i18n>Quotas</legend>
+
+          <!-- Max Bytes -->
+          <div class="form-group"
+               [ngClass]="{'has-error': form.showError('max_bytes', formDir)}">
+            <label class="control-label col-sm-3"
+                   for="max_bytes">
+              <ng-container i18n>Max bytes</ng-container>
+            </label>
+            <cd-helper>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: `cd-helper` should be inside the label, after the `ng-container`.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 11]
```diff
@@ -423,6 +423,61 @@
...
+          </div>
+
+          <!-- Max Objects -->
+          <div class="form-group"
+               [ngClass]="{'has-error': form.showError('max_objects', formDir)}">
+            <label class="control-label col-sm-3"
+                   for="max_objects">
+              <ng-container i18n>Max objects</ng-container>
+            </label>
+            <cd-helper>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: cd-helper should be inside the label, after the ng-container.

Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Thanks! done

**Comment by callithea** [comment]
jenkins test dashboard

**Comment by bk201** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
QA run failed:
http://pulpito.ceph.com/laura-2019-05-16_14:05:29-rados:mgr-wip-lpaduano-testing-27945-distro-basic-mira/
The failures are not related to the PR, the issue is already tracked here:
- https://tracker.ceph.com/issues/39648

Started another QA run on a different machine type:
http://pulpito.ceph.com/laura-2019-05-17_06:44:34-rados:mgr-wip-lpaduano-testing-27945-distro-basic-smithi/

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 12]
```diff
@@ -423,6 +423,61 @@
...
+          <!-- Max Bytes -->
+          <div class="form-group"
+               [ngClass]="{'has-error': form.showError('max_bytes', formDir)}">
+            <label class="control-label col-sm-3"
+                   for="max_bytes">
+              <ng-container i18n>Max bytes</ng-container>
+              <cd-helper>
+                <span i18n>Leave it blank or specify 0 to disable this quota.</span>
+                <br>
+                <span i18n>A valid quota should be greater then 0.</span>
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: "then" should be replaced with "than"

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 13]
```diff
@@ -423,6 +423,61 @@
...
+          <!-- Max Objects -->
+          <div class="form-group"
+               [ngClass]="{'has-error': form.showError('max_objects', formDir)}">
+            <label class="control-label col-sm-3"
+                   for="max_objects">
+              <ng-container i18n>Max objects</ng-container>
+              <cd-helper>
+                <span i18n>Leave it blank or specify 0 to disable this quota.</span>
+                <br>
+                <span i18n>A valid quota should be greater then 0.</span>
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: "then" should be replaced with "than".

Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Done, thanks!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 12 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html]: Done, thanks!

**Comment by tchaikov** [comment]
retest this please.

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
QA run was successful:
http://pulpito.ceph.com/laura-2019-05-23_07:20:49-rados:mgr-wip-lpaduano-testing-27945-28075-distro-basic-smithi/

**Comment by callithea** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
@LenzGr good to merge?

**Comment by LenzGr** [comment]
> @LenzGr good to merge?

Looks good, just needs a rebase to adapt to the latest changes to the Pools QA tests.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts [hunk 14]
```diff
@@ -529,7 +535,20 @@ export class PoolFormComponent implements OnInit {
...
             attr: 'name'
           },
-      { externalFieldName: 'rule_name', formControlName: 'crushRule', attr: 'rule_name' }
+      { externalFieldName: 'rule_name', formControlName: 'crushRule', attr: 'rule_name' },
+      {
+        externalFieldName: 'quota_max_bytes',
+        formControlName: 'max_bytes',
+        replaceFn: this.formatter.toBytes,
+        editable: true,
+        resetValue: this.editing ? 0 : undefined
```
Comment (bk201) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: When editing a pool, clearing quota field should disable pool quota.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts [hunk 15]
```diff
@@ -529,7 +535,20 @@ export class PoolFormComponent implements OnInit {
...
+      },
+      {
+        externalFieldName: 'quota_max_objects',
+        formControlName: 'max_objects',
+        editable: true,
+        resetValue: this.editing ? 0 : undefined
```
Comment (bk201) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: When editing a pool, clearing quota field should disable pool quota.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts (hunk 14 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts]: I've rebased and found previous changes missed this recommendation: https://github.com/ceph/ceph/pull/27945#discussion_r281976639

When editing a pool, clearing quota fields should disable corresponding quotas.
So I think it's better to start another round of review and test again, since pool testing code on master changed too.

**Comment by callithea** [comment]
QA run was successful:
http://pulpito.ceph.com/laura-2019-06-18_15:09:26-rados:mgr-wip-lpaduano-testing-27945-2-distro-basic-smithi/

**Comment by ricardoasmarques** [comment]
jenkins test make check arm64
